CDAP-5191 FTP Batch Source implementation

gokulavasan opened the pull request:
FTP Batch Source and Unit test

JIRA : https://issues.cask.co/browse/CDAP-5191

Commit:
Changing 'ts' to 'offset' for each line read from file source

Commit:
Merge branch 'develop' of github.com:caskdata/hydrator-plugins into develop

Commit:
Merge branch 'develop' of github.com:caskdata/hydrator-plugins into develop

gsps1 committed:
CDAP-5149 getSchema and listTables plugin endpoint implementations in DBSource

Commit:
Merge branch 'develop' of github.com:caskdata/hydrator-plugins into develop

albertshau committed:
CDAP-5146 add a group by aggregator

The plugin can group by fields in a record, then calculate
count, sum, min, max, avg, first, and last on the grouped values.
Aggregate functions should eventually be turned into plugins.

Commit:
CDAP-4856 Renamed TeraDataSource to DBSource and fixed all the tests

Commit:
CDAP-4856 Renamed the UI widgets

Commit:
CDAP-4856 Reordered imports

Commit:
CDAP-4856 Modified plugin name

gokulavasan committed:
CDAP-5191 FTP Batch Source implementation

Delete FTPInputStream, since it is present in CDAP AppFabric

Commit:
CDAP-4856 Addressing comments

Commit:
CDAP-4856 Removed ETLDBInputFormat and also fixed test cases

Commit:
Merge pull request #174 from caskdata/feature/teradata-and-db-merge

CDAP-4856 Replace DBSource with TeradataSource

gsps1 committed:
Merge pull request #172 from caskdata/feature/dbsource-getschema-list-tables-endpoints

getSchema and listTables plugin endpoint implementations in DBSource

gsps1 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FTPBatchSource.java:
```diff
@@ -0,0 +1,99 @@
+    } else {
+      providedProperties = GSON.fromJson(fsProperties, MAP_STRING_STRING_TYPE);
+    }
+    providedProperties.put(FileInputFormat.SPLIT_MINSIZE, Long.toString(Long.MAX_VALUE));
```
in the comment you say limiting the splits to 1, but the value is Long.MAX_VALUE, shouldn't it be 1 ?

gsps1 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FTPBatchSource.java:
```diff
@@ -0,0 +1,99 @@
+    public FTPBatchSourceConfig() {
+      this.fileSystemProperties = GSON.toJson(ImmutableMap.<String, String>of());
+      this.fileRegex = ".*";
+    }
```
can you extract these properties out of FileBatchConfig, so you don't have to pass null for those two params as well in constructor.

gsps1 commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/FileBatchSource.java:
```diff
@@ -88,7 +88,7 @@
     "If the field 'timeTable' is present, then it will read in files that have not yet been read. Defaults to '.*', " +
     "which indicates that no files will be filtered.";
-  private static final String FILESYSTEM_PROPERTIES_DESCRIPTION = "A JSON string representing a map of properties " +
+  protected static final String FILESYSTEM_PROPERTIES_DESCRIPTION = "A JSON string representing a map of properties " +
```
why this change?

gokulavasan replied:
So that we can access the property description in FTPBatchSourceConfig.

gsps1 commented on core-plugins/src/test/java/co/cask/hydrator/plugin/batch/ETLFTPTestRun.java:
```diff
@@ -0,0 +1,124 @@
+  private static final String TEST_STRING = "Hello World";
+
+  @ClassRule
+  public static final TemporaryFolder TMP_FOLDER = new TemporaryFolder();
```
can you reuse the Temp folder in test base ?

gokulavasan replied:
Yes I can. Will remove.

albertshau committed:
make overflow less likely for average and some small cleanup

gsps1 commented on core-plugins/src/test/java/co/cask/hydrator/plugin/batch/ETLFTPTestRun.java:
```diff
@@ -0,0 +1,124 @@
+    ftpServer.addUserAccount(new UserAccount(USER, PWD, folder.getAbsolutePath()));
+    ftpServer.start();
+
+    Assert.assertTrue(ftpServer.isStarted());
```
can this be false? if so, should we retry this check for it to be started

gokulavasan replied:
Yes, I will add a try mechanism to check

gsps1 commented:
what would happen in failure scenario's ?  path is incorrect, access errors, etc? would we return appropriate error messages?

albertshau committed:
check for null before adding schema field

albertshau committed:
allow each module to specify their own parents

Since the data-pipeline app supports some plugin types that
the etl-batch app does not, we need a way for certain modules to
limit which apps they extend.

albertshau committed:
Merge pull request #176 from caskdata/group-by-aggregator

CDAP-5146 add a group by aggregator

albertshau committed:
Merge pull request #178 from caskdata/per-module-parents

allow each module to specify their own parents

gokulavasan committed:
Addressing conflicts

gokulavasan commented:
@gsps1 Regarding the question about failure scenarios, we delegate to FileBatchSource. We don't perform any additional checks.

gokulavasan committed:
Address PR comments and ignore test

gokulavasan committed:
Addressing merge conflicts

gokulavasan commented:
Reopened here https://github.com/caskdata/hydrator-plugins/pull/182
